pkg/pool-hooks/contracts/StableSurgeHook.sol

```diff
@@ -0,0 +1,338 @@
+        address pool,
+        uint256 newMaxSurgeSurgeFeePercentage
+    ) external withValidPercentage(newMaxSurgeSurgeFeePercentage) withPermission(pool) {
+        _setMaxSurgeFeePercentage(pool, newMaxSurgeSurgeFeePercentage);
+    }
+
+    /**
+     * @notice Sets the hook threshold percentage.
+     * @dev This function must be permissioned. If the pool does not have a swap fee manager role set, the surge
+     * threshold can only be changed by governance. It is initially set to the default threshold for this hook contract.
+     */
+    function setSurgeThresholdPercentage(
+        address pool,
+        uint256 newSurgeThresholdPercentage
+    ) external withValidPercentage(newSurgeThresholdPercentage) withPermission(pool) {
+        _setSurgeThresholdPercentage(pool, newSurgeThresholdPercentage);
+    }
+
+    /// @dev Ensure the sender is the swapFeeManager, or default to governance if there is no manager.
+    function _ensureValidSender(address pool) private view {
+        address swapFeeManager = _vault.getPoolRoleAccounts(pool).swapFeeManager;
+
+        if (swapFeeManager == address(0)) {
+            if (_canPerform(getActionId(msg.sig), msg.sender, pool) == false) {
+                revert SenderNotAllowed();
+            }
+        } else if (swapFeeManager != msg.sender) {
+            revert SenderNotAllowed();
+        }
+    }
+
+    /**
+     * @notice Calculate the surge fee percentage. If below threshold, return the standard static swap fee percentage.
+     * @dev It is public to allow it to be called off-chain.
+     * @param params Input parameters for the swap (balances needed)
+     * @param pool The pool we are computing the fee for
+     * @param staticFeePercentage The static fee percentage for the pool (default if there is no surge)
+     */
+    function getSurgeFeePercentage(
+        PoolSwapParams calldata params,
+        address pool,
+        uint256 staticFeePercentage
+    ) public view returns (uint256) {
+        uint256 numTokens = params.balancesScaled18.length;
+
+        uint256 invariant = IBasePool(pool).computeInvariant(params.balancesScaled18, Rounding.ROUND_DOWN);
+        (uint256 currentAmp, , ) = IStablePool(pool).getAmplificationParameter();
+
+        uint256 amountCalculatedScaled18;
+        if (params.kind == SwapKind.EXACT_IN) {
+            amountCalculatedScaled18 = StableMath.computeOutGivenExactIn(
+                currentAmp,
+                params.balancesScaled18,
+                params.indexIn,
+                params.indexOut,
+                params.amountGivenScaled18,
+                invariant
+            );
+        } else {
+            amountCalculatedScaled18 = StableMath.computeInGivenExactOut(
+                currentAmp,
+                params.balancesScaled18,
+                params.indexIn,
+                params.indexOut,
+                params.amountGivenScaled18,
+                invariant
+            );
+        }
```

Comment:
I'd just call `onSwap`, removing `onlyVault` tbh. At the end of the day it's a `view` function in this case; I don't see how calling it can hurt anyone. If it wasn't `view` it'd be a different story. I don't think you can force any sort of read only reentrancy in this case. The code here should do exactly the same as all the inputs to `onSwap` should remain the same (amp, balances, index in / out); if you think I'm mistaken here please correct me :)

That won't work with the current deployed factory but I guess that's ok. I think I prefer that as opposed to copying the code here.